Add realm.inWriteTransaction()

```
Caused by: java.lang.IllegalStateException: Cannot cancel a non-write transaction.
      at io.realm.internal.ImplicitTransaction.rollbackAndContinueAsRead(ImplicitTransaction.java:60)
      at io.realm.internal.SharedGroupManager.rollbackAndContinueAsRead(SharedGroupManager.java:110)
      at io.realm.BaseRealm.cancelTransaction(BaseRealm.java:329)
      at io.realm.Realm.cancelTransaction(Realm.java:112)
      at com.couchgram.privacycall.utils.CallStickerDecompress.initZip(CallStickerDecompress.java:122)
      at com.couchgram.privacycall.utils.CallStickerDecompress.<init>(CallStickerDecompress.java:56)
      at com.couchgram.privacycall.utils.Utils.setDefaultCallStickerResource(Utils.java:1526)
      at com.couchgram.privacycall.activity.GuideActivity.lambda$onCreate$12(GuideActivity.java:247)
      at com.couchgram.privacycall.activity.GuideActivity.access$lambda$0(GuideActivity.java)
      at com.couchgram.privacycall.activity.GuideActivity$$Lambda$1.call()
      at rx.Observable.unsafeSubscribe(Observable.java:7710)
      at rx.internal.operators.OperatorSubscribeOn$1$1.call(OperatorSubscribeOn.java:62)
      at rx.internal.schedulers.ScheduledAction.run(ScheduledAction.java:55)
      at java.util.concurrent.Executors$RunnableAdapter.call(Executors.java:442)
      at java.util.concurrent.FutureTask$Sync.innerRun(FutureTask.java:305)
      at java.util.concurrent.FutureTask.run(FutureTask.java:137)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.access$201(ScheduledThreadPoolExecutor.java:150)
      at java.util.concurrent.ScheduledThreadPoolExecutor$ScheduledFutureTask.run(ScheduledThreadPoolExecutor.java:264)
      at java.util.concurrent.ThreadPoolExecutor.runWorker(ThreadPoolExecutor.java:1076)
      at java.util.concurrent.ThreadPoolExecutor$Worker.run(ThreadPoolExecutor.java:569)
      at java.lang.Thread.run(Thread.java:856)
```

```
Observable.create(subscriber -> {
           Utils.setDefaultCallStickerResource(getApplicationContext(), true);
           subscriber.onCompleted();
       }).subscribeOn(Schedulers.io()).observeOn(Schedulers.newThread()).subscribe();
```

Developer who reported this: @chonamdoo

--- a/changelog.txt
+++ b/changelog.txt
@@ -5,6 +5,7 @@
  * Added Realm.isClosed() method.
  * Added Realm.distinct() method.
  * Added RealmQuery.isValid(), RealmResults.isValid() and RealmList.isValid(). Each method checks whether the instance is still valid to use or not(for example, the Realm has been closed or any parent object has been removed).
+ * Added Realm.isInWriteTransaction() method.
  * Following methods in managed RealmList now throw IllegalStateException instead of native crash when RealmList.isValid() returns false: add(int,RealmObject), add(RealmObject)
  * Following methods in managed RealmList now throw IllegalStateException instead of ArrayIndexOutOfBoundsException when RealmList.isValid() returns false: set(int,RealmObject), move(int,int), remove(int), get(int)
  * Following methods in managed RealmList now throw IllegalStateException instead of returning 0/null when RealmList.isValid() returns false: clear(), removeAll(Collection), remove(RealmObject), first(), last(), size(), where()

--- a/changelog.txt
+++ b/changelog.txt
@@ -5,7 +5,7 @@
  * Added Realm.isClosed() method.
  * Added Realm.distinct() method.
  * Added RealmQuery.isValid(), RealmResults.isValid() and RealmList.isValid(). Each method checks whether the instance is still valid to use or not(for example, the Realm has been closed or any parent object has been removed).
- * Added Realm.isInWriteTransaction() method.
+ * Added Realm.isInTransaction() method.
  * Following methods in managed RealmList now throw IllegalStateException instead of native crash when RealmList.isValid() returns false: add(int,RealmObject), add(RealmObject)
  * Following methods in managed RealmList now throw IllegalStateException instead of ArrayIndexOutOfBoundsException when RealmList.isValid() returns false: set(int,RealmObject), move(int,int), remove(int), get(int)
  * Following methods in managed RealmList now throw IllegalStateException instead of returning 0/null when RealmList.isValid() returns false: clear(), removeAll(Collection), remove(RealmObject), first(), last(), size(), where()
@@ -28,7 +28,7 @@
  * Deprecated methods: RealmQuery.minimum{Int,Float,Double}, RealmQuery.maximum{Int,Float,Double}. Use RealmQuery.min() and RealmQuery.max() instead.
  * Added support for x86_64.
  * Fixed an issue where opening the same Realm file on two Looper threads could potentially lead to an IllegalStateException being thrown.
- * Opening a Realm file from one thread will no longer be blocked by a write transaction from another thread.
+ * Opening a Realm file from one thread will no longer be blocked by a transaction from another thread.
  * Fixed an issue preventing the call of listeners on refresh().
  * Range restrictions of Date fields have been removed. Date fields now accepts any value. Milliseconds are still removed.
 
@@ -230,7 +230,7 @@
   * Better support for Java 7 and 8 in the annotations processor
   * Better support for the Eclipse annotations processor
   * Added Eclipse support to the distribution folder
-  * Added Realm.cancelTransaction() to cancel/abort/rollback a write transaction
+  * Added Realm.cancelTransaction() to cancel/abort/rollback a transaction
   * Faster implementation of RealmQuery.findFirst()
   * Upgraded core to 0.85.1 (deep copying of strings in queries; preparation for link queries)
   * Added support for link queries in the form realm.where(Owner.class).equalTo("cat.age", 12).findAll()
@@ -239,7 +239,7 @@
   * Simplified the release artifact to a single Jar file
   * Added support for Eclipse
   * Added support for deploying to Maven
-  * Throw exception if nested write transactions are used (it's not allowed)
+  * Throw exception if nested transactions are used (it's not allowed)
   * Javadoc updated
   * Fixed bug in RealmResults: https://github.com/realm/realm-java/issues/453
   * New annotation @Index to add search index to a field (currently only supporting String fields)

--- a/examples/introExample/src/main/java/io/realm/examples/intro/IntroExampleActivity.java
+++ b/examples/introExample/src/main/java/io/realm/examples/intro/IntroExampleActivity.java
@@ -96,14 +96,14 @@ public class IntroExampleActivity extends Activity {
         person.setName("Young Person");
         person.setAge(14);
 
-        // When the write transaction is committed, all changes a synced to disk.
+        // When the transaction is committed, all changes a synced to disk.
         realm.commitTransaction();
 
         // Find the first person (no query conditions) and read a field
         person = realm.where(Person.class).findFirst();
         showStatus(person.getName() + ":" + person.getAge());
 
-        // Update person in a write transaction
+        // Update person in a transaction
         realm.beginTransaction();
         person.setName("Senior Person");
         person.setAge(99);
@@ -141,7 +141,7 @@ public class IntroExampleActivity extends Activity {
         // Those can not be transferred across threads.
         Realm realm = Realm.getInstance(this);
 
-        // Add ten persons in one write transaction
+        // Add ten persons in one transaction
         realm.beginTransaction();
         Dog fido = realm.createObject(Dog.class);
         fido.setName("fido");

--- a/examples/kotlinExample/src/main/java/io/realm/examples/kotlin/KotlinExampleActivity.kt
+++ b/examples/kotlinExample/src/main/java/io/realm/examples/kotlin/KotlinExampleActivity.kt
@@ -93,14 +93,14 @@ public class KotlinExampleActivity : Activity() {
         person.name = "Young Person"
         person.age = 14
 
-        // When the write transaction is committed, all changes a synced to disk.
+        // When the transaction is committed, all changes a synced to disk.
         realm.commitTransaction()
 
         // Find the first person (no query conditions) and read a field
         person = realm.where(javaClass<Person>()).findFirst()
         showStatus(person.name + ": " + person.age)
 
-        // Update person in a write transaction
+        // Update person in a transaction
         realm.beginTransaction()
         person.name = "Senior Person"
         person.age = 99
@@ -140,7 +140,7 @@ public class KotlinExampleActivity : Activity() {
         // Those can not be transferred across threads.
         val realm = Realm.getInstance(this)
 
-        // Add ten persons in one write transaction
+        // Add ten persons in one transaction
         realm.beginTransaction()
         val fido = realm.createObject(javaClass<Dog>())
         fido.name = "fido"

--- a/realm/src/androidTest/java/io/realm/RealmAsyncQueryTests.java
+++ b/realm/src/androidTest/java/io/realm/RealmAsyncQueryTests.java
@@ -42,12 +42,12 @@ import io.realm.instrumentation.MockActivityManager;
 import io.realm.proxy.HandlerProxy;
 
 public class RealmAsyncQueryTests extends InstrumentationTestCase {
-    // **********************************
-    // ****  Async write transaction  ***
-    // **********************************
+    // ****************************
+    // ****  Async transaction  ***
+    // ****************************
 
-    // start asynchronously a write transaction to insert one element
-    public void testAsyncWriteTransaction() throws Throwable {
+    // start asynchronously a transaction to insert one element
+    public void testAsyncTransaction() throws Throwable {
         final CountDownLatch signalCallbackFinished = new CountDownLatch(1);
         final Realm[] realm = new Realm[1];
         final Throwable[] threadAssertionError = new Throwable[1];// to catch both Exception & AssertionError
@@ -60,7 +60,7 @@ public class RealmAsyncQueryTests extends InstrumentationTestCase {
                 backgroundLooper[0] = Looper.myLooper();
 
                 try {
-                    realm[0] = openRealmInstance("testAsyncWriteTransaction");
+                    realm[0] = openRealmInstance("testAsyncTransaction");
 
                     assertEquals(0, realm[0].allObjects(Owner.class).size());
 

--- a/realm/src/androidTest/java/io/realm/RealmListTest.java
+++ b/realm/src/androidTest/java/io/realm/RealmListTest.java
@@ -711,8 +711,8 @@ public class RealmListTest extends AndroidTestCase {
         testRealm.commitTransaction();
     }
 
-    // Test that all methods that require a write transaction (ie. any function that mutates Realm data)
-    public void testMutableMethodsOutsideWriteTransactions() {
+    // Test that all methods that require a transaction (ie. any function that mutates Realm data)
+    public void testMutableMethodsOutsideTransactions() {
         testRealm.beginTransaction();
         RealmList<Dog> list = testRealm.createObject(AllTypes.class).getColumnRealmList();
         Dog dog = testRealm.createObject(Dog.class);

--- a/realm/src/androidTest/java/io/realm/RealmResultsTest.java
+++ b/realm/src/androidTest/java/io/realm/RealmResultsTest.java
@@ -1078,8 +1078,8 @@ public class RealmResultsTest extends AndroidTestCase {
     }
 
 
-    // Test that all methods that require a write transaction (ie. any function that mutates Realm data)
-    public void testMutableMethodsOutsideWriteTransactions() {
+    // Test that all methods that require a transaction (ie. any function that mutates Realm data)
+    public void testMutableMethodsOutsideTransactions() {
         RealmResults<AllTypes> result = testRealm.where(AllTypes.class).findAll();
 
         try {

--- a/realm/src/androidTest/java/io/realm/RealmTest.java
+++ b/realm/src/androidTest/java/io/realm/RealmTest.java
@@ -666,6 +666,15 @@ public class RealmTest extends AndroidTestCase {
         assertEquals(TEST_DATA_SIZE + 1, resultList.size());
     }
 
+    public void testCommitTransactionAfterCancelTransaction () {
+        testRealm.beginTransaction();
+        testRealm.cancelTransaction();
+        try {
+            testRealm.commitTransaction();
+            fail();
+        } catch (IllegalStateException ignore) {
+        }
+    }
 
     public void testCancelTransaction() {
         populateTestRealm();
@@ -2160,5 +2169,17 @@ public class RealmTest extends AndroidTestCase {
             } catch (UnsupportedOperationException ignore) {
             }
         }
+    }
+
+    public void testIsInWriteTransaction() {
+        assertFalse(testRealm.isInWriteTransaction());
+        testRealm.beginTransaction();
+        assertTrue(testRealm.isInWriteTransaction());
+        testRealm.commitTransaction();
+        assertFalse(testRealm.isInWriteTransaction());
+        testRealm.beginTransaction();
+        assertTrue(testRealm.isInWriteTransaction());
+        testRealm.cancelTransaction();
+        assertFalse(testRealm.isInWriteTransaction());
     }
 }

--- a/realm/src/androidTest/java/io/realm/RealmTest.java
+++ b/realm/src/androidTest/java/io/realm/RealmTest.java
@@ -16,7 +16,6 @@
 package io.realm;
 
 import android.content.Context;
-import android.os.Handler;
 import android.test.AndroidTestCase;
 
 import junit.framework.AssertionFailedError;
@@ -32,7 +31,6 @@ import java.util.ArrayList;
 import java.util.Arrays;
 import java.util.Date;
 import java.util.List;
-import java.util.Map;
 import java.util.Random;
 import java.util.Scanner;
 import java.util.concurrent.Callable;
@@ -769,12 +767,12 @@ public class RealmTest extends AndroidTestCase {
         // These calls should fail outside a Transaction:
         try {
             testRealm.createObject(AllTypes.class);
-            fail("Realm.createObject should fail outside write transaction");
+            fail("Realm.createObject should fail outside transaction");
         } catch (IllegalStateException ignored) {
         }
         try {
             testRealm.remove(AllTypes.class, 0);
-            fail("Realm.remove should fail outside write transaction");
+            fail("Realm.remove should fail outside transaction");
         } catch (IllegalStateException ignored) {
         }
     }
@@ -1708,8 +1706,8 @@ public class RealmTest extends AndroidTestCase {
         assertTrue(tmpFile.createNewFile());
     }
 
-    // Test that all methods that require a write transaction (ie. any function that mutates Realm data)
-    public void testMutableMethodsOutsideWriteTransactions() throws JSONException, IOException {
+    // Test that all methods that require a transaction (ie. any function that mutates Realm data)
+    public void testMutableMethodsOutsideTransactions() throws JSONException, IOException {
 
         // Prepare standalone object data
         AllTypesPrimaryKey t = new AllTypesPrimaryKey();
@@ -1726,7 +1724,7 @@ public class RealmTest extends AndroidTestCase {
         InputStream jsonArrStream = TestHelper.stringToStream(jsonArrStr);
         InputStream jsonArrStream2 = TestHelper.stringToStream(jsonArrStr);
 
-        // Test all methods that should require a write transaction
+        // Test all methods that should require a transaction
         try { testRealm.createObject(AllTypes.class);   fail(); } catch (IllegalStateException expected) {}
         try { testRealm.copyToRealm(t);                 fail(); } catch (IllegalStateException expected) {}
         try { testRealm.copyToRealm(ts);                fail(); } catch (IllegalStateException expected) {}
@@ -2171,15 +2169,15 @@ public class RealmTest extends AndroidTestCase {
         }
     }
 
-    public void testIsInWriteTransaction() {
-        assertFalse(testRealm.isInWriteTransaction());
+    public void testIsInTransaction() {
+        assertFalse(testRealm.isInTransaction());
         testRealm.beginTransaction();
-        assertTrue(testRealm.isInWriteTransaction());
+        assertTrue(testRealm.isInTransaction());
         testRealm.commitTransaction();
-        assertFalse(testRealm.isInWriteTransaction());
+        assertFalse(testRealm.isInTransaction());
         testRealm.beginTransaction();
-        assertTrue(testRealm.isInWriteTransaction());
+        assertTrue(testRealm.isInTransaction());
         testRealm.cancelTransaction();
-        assertFalse(testRealm.isInWriteTransaction());
+        assertFalse(testRealm.isInTransaction());
     }
 }

--- a/realm/src/main/java/io/realm/BaseRealm.java
+++ b/realm/src/main/java/io/realm/BaseRealm.java
@@ -78,7 +78,7 @@ abstract class BaseRealm implements Closeable {
     // List of Realm files that has already been validated
     static final Set<String> validatedRealmFiles = new HashSet<String>();
 
-    // thread pool for all async operations (Query & Write transaction)
+    // thread pool for all async operations (Query & transaction)
     static final RealmThreadPoolExecutor asyncQueryExecutor = RealmThreadPoolExecutor.getInstance();
 
     protected final List<WeakReference<RealmChangeListener>> changeListeners =
@@ -110,7 +110,6 @@ abstract class BaseRealm implements Closeable {
      * This feature is only available if the Realm instance lives is a {@link android.os.Looper} enabled thread.
      *
      * @param autoRefresh true will turn auto-refresh on, false will turn it off.
-     * @throws java.lang.IllegalStateException if trying to enable auto-refresh in a thread without Looper.
      */
     public void setAutoRefresh(boolean autoRefresh) {
         checkIfValid();
@@ -137,12 +136,10 @@ abstract class BaseRealm implements Closeable {
     }
 
     /**
-     * Indicates if the Realm is currently in a write transaction.
-     * @return the write transaction status.
-     * @throws java.lang.IllegalStateException if the Realm is not opened or in a different thread
-     *                                      than the one it was created on.
+     * Checks if the Realm is currently in a transaction.
+     * @return {@code true} if inside a transaction, {@code false} otherwise.
      */
-    public boolean isInWriteTransaction() {
+    public boolean isInTransaction() {
         checkIfValid();
         return !sharedGroupManager.isImmutable();
     }
@@ -253,8 +250,8 @@ abstract class BaseRealm implements Closeable {
      * <p>
      * The destination file cannot already exist.
      * <p>
-     * Note that if this is called from within a write transaction it writes the
-     * current data, and not the data as it was when the last write transaction was committed.
+     * Note that if this is called from within a transaction it writes the
+     * current data, and not the data as it was when the last transaction was committed.
      *
      * @param destination File to save the Realm to
      * @throws java.io.IOException if any write operation fails
@@ -268,8 +265,8 @@ abstract class BaseRealm implements Closeable {
      * <p>
      * The destination file cannot already exist.
      * <p>
-     * Note that if this is called from within a write transaction it writes the
-     * current data, and not the data as it was when the last write transaction was committed.
+     * Note that if this is called from within a transaction it writes the current data, and not the
+     * data as it was when the last transaction was committed.
      * <p>
      * @param destination File to save the Realm to
      * @param key a 64-byte encryption key
@@ -296,18 +293,15 @@ abstract class BaseRealm implements Closeable {
     }
 
     /**
-     * Starts a write transaction, this must be closed with {@link io.realm.Realm#commitTransaction()}
-     * or aborted by {@link io.realm.Realm#cancelTransaction()}. Write transactions are used to
-     * atomically create, update and delete objects within a realm.
+     * Starts a transaction, this must be closed with {@link io.realm.Realm#commitTransaction()} or
+     * aborted by {@link io.realm.Realm#cancelTransaction()}. Transactions are used to atomically
+     * create, update and delete objects within a realm.
      * <br>
-     * Before beginning the write transaction, {@link io.realm.Realm#beginTransaction()} updates the
-     * realm in the case of pending updates from other threads.
+     * Before beginning the transaction, {@link io.realm.Realm#beginTransaction()} updates the realm
+     * in the case of pending updates from other threads.
      * <br>
-     * Notice: it is not possible to nest write transactions. If you start a write
-     * transaction within a write transaction an exception is thrown.
-     * <br>
-     * @throws java.lang.IllegalStateException If already in a write transaction or incorrect thread.
-     *
+     * Notice: it is not possible to nest transactions. If you start a transaction within a
+     * transaction an exception is thrown.
      */
     public void beginTransaction() {
         checkIfValid();
@@ -320,9 +314,6 @@ abstract class BaseRealm implements Closeable {
      * that a change has occurred. When the event is received, the other Realms will get their
      * objects and {@link io.realm.RealmResults} updated to reflect
      * the changes from this commit.
-     *
-     * @throws java.lang.IllegalStateException if the write transaction is in an invalid state
-     *                                          or incorrect thread.
      */
     public void commitTransaction() {
         checkIfValid();
@@ -361,10 +352,7 @@ abstract class BaseRealm implements Closeable {
      * <br>
      * The Realm reverts back to read-only.
      * <br>
-     * Calling this when not in a write transaction will throw an exception.
-     *
-     * @throws java.lang.IllegalStateException    If the write transaction is an invalid state,
-     *                                             not in a write transaction or incorrect thread.
+     * Calling this when not in a transaction will throw an exception.
      */
     public void cancelTransaction() {
         checkIfValid();
@@ -422,9 +410,6 @@ abstract class BaseRealm implements Closeable {
      * <p>
      * It's important to always remember to close Realm instances when you're done with it in order
      * not to leak memory, file descriptors or grow the size of Realm file out of measure.
-     *
-     * @throws java.lang.IllegalStateException if trying to close Realm on a different thread than the
-     * one it was created on.
      */
     @Override
     public void close() {

--- a/realm/src/main/java/io/realm/BaseRealm.java
+++ b/realm/src/main/java/io/realm/BaseRealm.java
@@ -137,6 +137,17 @@ abstract class BaseRealm implements Closeable {
     }
 
     /**
+     * Indicates if the Realm is currently in a write transaction.
+     * @return the write transaction status.
+     * @throws java.lang.IllegalStateException if the Realm is not opened or in a different thread
+     *                                      than the one it was created on.
+     */
+    public boolean isInWriteTransaction() {
+        checkIfValid();
+        return !sharedGroupManager.isImmutable();
+    }
+
+    /**
      * Add a change listener to the Realm.
      * <p>
      * The listeners will be executed:
@@ -310,7 +321,7 @@ abstract class BaseRealm implements Closeable {
      * objects and {@link io.realm.RealmResults} updated to reflect
      * the changes from this commit.
      *
-     * @throws java.lang.IllegalStateException If the write transaction is in an invalid state
+     * @throws java.lang.IllegalStateException if the write transaction is in an invalid state
      *                                          or incorrect thread.
      */
     public void commitTransaction() {

--- a/realm/src/main/java/io/realm/BaseRealm.java
+++ b/realm/src/main/java/io/realm/BaseRealm.java
@@ -310,7 +310,8 @@ abstract class BaseRealm implements Closeable {
      * objects and {@link io.realm.RealmResults} updated to reflect
      * the changes from this commit.
      *
-     * @throws java.lang.IllegalStateException If the write transaction is in an invalid state or incorrect thread.
+     * @throws java.lang.IllegalStateException If the write transaction is in an invalid state
+     *                                          or incorrect thread.
      */
     public void commitTransaction() {
         checkIfValid();

--- a/realm/src/main/java/io/realm/Realm.java
+++ b/realm/src/main/java/io/realm/Realm.java
@@ -1157,8 +1157,6 @@ public final class Realm extends BaseRealm {
      *
      * @param configuration A {@link RealmConfiguration}
      * @return false if a file could not be deleted. The failing file will be logged.
-     *
-     * @throws java.lang.IllegalStateException if trying to delete a Realm that is already open.
      */
     public static boolean deleteRealm(RealmConfiguration configuration) {
         return BaseRealm.deleteRealm(configuration);
@@ -1175,8 +1173,6 @@ public final class Realm extends BaseRealm {
      *
      * @param configuration a {@link RealmConfiguration} pointing to a Realm file.
      * @return true if successful, false if any file operation failed
-     *
-     * @throws java.lang.IllegalStateException if trying to compact a Realm that is already open.
      */
     public static boolean compactRealm(RealmConfiguration configuration) {
         return BaseRealm.compactRealm(configuration);

--- a/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
+++ b/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
@@ -44,22 +44,20 @@ public class ImplicitTransaction extends Group {
 
     public void promoteToWrite() {
         assertNotClosed();
-        if (immutable) {
-            immutable = false;
-            parent.promoteToWrite();
-        } else {
+        if (!immutable) {
             throw new IllegalStateException("Nested transactions are not allowed. Use commitTransaction() after each beginTransaction().");
         }
+        immutable = false;
+        parent.promoteToWrite();
     }
 
     public void commitAndContinueAsRead() {
         assertNotClosed();
-        if (!immutable) {
-            parent.commitAndContinueAsRead();
-            immutable = true;
-        } else {
-            throw new IllegalStateException("Cannot commit a non-write transaction.");
+        if (immutable) {
+            throw new IllegalStateException("Cannot commit a non transaction.");
         }
+        parent.commitAndContinueAsRead();
+        immutable = true;
     }
 
     public void endRead() {
@@ -69,12 +67,11 @@ public class ImplicitTransaction extends Group {
 
     public void rollbackAndContinueAsRead() {
         assertNotClosed();
-        if (!immutable) {
-            parent.rollbackAndContinueAsRead();
-            immutable = true;
-        } else {
-            throw new IllegalStateException("Cannot cancel a non-write transaction.");
+        if (immutable) {
+            throw new IllegalStateException("Cannot cancel a non transaction.");
         }
+        parent.rollbackAndContinueAsRead();
+        immutable = true;
     }
 
     private void assertNotClosed() {

--- a/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
+++ b/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
@@ -54,7 +54,7 @@ public class ImplicitTransaction extends Group {
     public void commitAndContinueAsRead() {
         assertNotClosed();
         if (immutable) {
-            throw new IllegalStateException("Cannot commit a non transaction.");
+            throw new IllegalStateException("Not inside a transaction.");
         }
         parent.commitAndContinueAsRead();
         immutable = true;
@@ -68,7 +68,7 @@ public class ImplicitTransaction extends Group {
     public void rollbackAndContinueAsRead() {
         assertNotClosed();
         if (immutable) {
-            throw new IllegalStateException("Cannot cancel a non transaction.");
+            throw new IllegalStateException("Not inside a transaction.");
         }
         parent.rollbackAndContinueAsRead();
         immutable = true;

--- a/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
+++ b/realm/src/main/java/io/realm/internal/ImplicitTransaction.java
@@ -54,8 +54,12 @@ public class ImplicitTransaction extends Group {
 
     public void commitAndContinueAsRead() {
         assertNotClosed();
-        parent.commitAndContinueAsRead();
-        immutable = true;
+        if (!immutable) {
+            parent.commitAndContinueAsRead();
+            immutable = true;
+        } else {
+            throw new IllegalStateException("Cannot commit a non-write transaction.");
+        }
     }
 
     public void endRead() {

--- a/realm/src/main/java/io/realm/internal/SharedGroupManager.java
+++ b/realm/src/main/java/io/realm/internal/SharedGroupManager.java
@@ -41,7 +41,7 @@ public class SharedGroupManager implements Closeable {
     /**
      * Creates a new instance of the FileWrapper for the given configuration on this thread.
      */
-    public  SharedGroupManager(RealmConfiguration configuration) {
+    public SharedGroupManager(RealmConfiguration configuration) {
         this.sharedGroup = new SharedGroup(
                 configuration.getPath(),
                 SharedGroup.IMPLICIT_TRANSACTION,

--- a/realm/src/main/java/io/realm/internal/SharedGroupManager.java
+++ b/realm/src/main/java/io/realm/internal/SharedGroupManager.java
@@ -155,7 +155,7 @@ public class SharedGroupManager implements Closeable {
     }
 
     /**
-     * Returns if the Realm is currently not in a write transaction.
+     * Returns if the Realm is currently not in a transaction.
      */
     public boolean isImmutable() {
         return transaction.immutable;

--- a/realm/src/main/java/io/realm/internal/SharedGroupManager.java
+++ b/realm/src/main/java/io/realm/internal/SharedGroupManager.java
@@ -155,6 +155,13 @@ public class SharedGroupManager implements Closeable {
     }
 
     /**
+     * Returns if the Realm is currently not in a write transaction.
+     */
+    public boolean isImmutable() {
+        return transaction.immutable;
+    }
+
+    /**
      * Compacts a Realm file. It cannot be open when calling this method.
      */
     public static boolean compact(RealmConfiguration configuration) {

